File: src/whombat/routes/user_runs.py
"""REST API routes for model runs."""
from uuid import UUID

from fastapi import APIRouter, Depends, UploadFile
from soundevent.io import aoef

from whombat import api, schemas
from whombat.dependencies import Session, ActiveUser
from whombat.filters.user_runs import UserRunFilter
from whombat.routes.types import Limit, Offset

__all__ = [
    "user_runs_router",
]


user_runs_router = APIRouter()


@user_runs_router.get("/", response_model=list[schemas.UserRun])
async def get_user_runs(
    session: Session,
    limit: Limit = 100,
    offset: Offset = 0,
    filter: UserRunFilter = Depends(UserRunFilter),  # type: ignore
) -> schemas.Page[schemas.UserRun]:
    """Get list of model runs."""
    user_runs, total = await api.user_runs.get_many(
        session,
        limit=limit,
        offset=offset,
        filters=[filter],
    )
    return schemas.Page(
        items=user_runs,
        total=total,
        offset=offset,
        limit=limit,
    )


@user_runs_router.post("/", response_model=schemas.UserRun)
async def create_user_run(
    session: Session,
    user: ActiveUser,
) -> schemas.UserRun:
    """Create model run."""
    user_run = await api.user_runs.create(
        session,
        user=user,
    )
    await session.commit()
    return user_run


@user_runs_router.get("/detail/", response_model=schemas.UserRun)
async def get_user_run(
    session: Session,
    user_run_uuid: UUID,
) -> schemas.UserRun:
    """Get model run."""
    return await api.user_runs.get(session, user_run_uuid)


@user_runs_router.put("/detail/", response_model=schemas.UserRun)
async def update_user_run(
    session: Session,
    user_run_uuid: UUID,
    data: schemas.UserRunUpdate,
) -> schemas.UserRun:
    """Update model run."""
    user_run = await api.user_runs.get(session, user_run_uuid)
    user_run = await api.user_runs.update(
        session,
        user_run,
        data,
    )
    await session.commit()
    return user_run


@user_runs_router.delete("/detail/", response_model=schemas.UserRun)
async def delete_user_run(
    session: Session,
    user_run_uuid: UUID,
) -> schemas.UserRun:
    """Delete model run."""
    user_run = await api.user_runs.get(session, user_run_uuid)
    await api.user_runs.delete(session, user_run)
    await session.commit()
    return user_run


@user_runs_router.get("/import//", response_model=schemas.UserRun)
async def import_user_run(
    session: Session,
    upload_file: UploadFile,
) -> schemas.UserRun:
    """Import model run."""
    obj = aoef.AOEFObject.model_validate_json(upload_file.file.read())
    data = aoef.to_soundevent(obj)
    user_run = await api.user_runs.from_soundevent(
        session,
        data,  # type: ignore
    )
    await session.commit()
    return user_run

File: front/src/lib/api/user_runs.ts
import { AxiosInstance } from "axios";

import { GetMany, Page } from "@/lib/api/common";
import * as schemas from "@/lib/schemas";
import type * as types from "@/lib/types";

const DEFAULT_ENDPOINTS = {
  getMany: "/api/v1/user_runs/",
  get: "/api/v1/user_runs/detail/",
  create: "/api/v1/user_runs/",
  delete: "/api/v1/user_runs/detail/",
};

export function registerUserRunAPI(
  instance: AxiosInstance,
  endpoints: typeof DEFAULT_ENDPOINTS = DEFAULT_ENDPOINTS,
) {
  async function createUserRun(): Promise<types.UserRun> {
    const response = await instance.post(endpoints.create);
    return schemas.UserRunSchema.parse(response.data);
  }

  async function getManyUserRuns(
    query: types.GetMany & types.UserRunFilter,
  ): Promise<types.Page<types.UserRun>> {
    const params = GetMany(schemas.UserRunFilterSchema).parse(query);
    const response = await instance.get(endpoints.getMany, {
      params: {
        limit: params.limit,
        offset: params.offset,
        sort_by: params.sort_by,
        user__eq: params.user?.id,
        evaluated__eq: params.evaluated,
        score__lt: params.score?.lt,
        score__gt: params.score?.gt,
        evaluation_set__eq: params.evaluation_set?.uuid,
        has_evaluation__eq: params.has_evaluation,
      },
    });
    return Page(schemas.UserRunSchema).parse(response.data);
  }

  async function getUserRun(uuid: string): Promise<types.UserRun> {
    const response = await instance.get(endpoints.get, {
      params: { user_run_uuid: uuid },
    });
    return schemas.UserRunSchema.parse(response.data);
  }

  async function deleteUserRun(userRun: types.UserRun): Promise<types.UserRun> {
    const response = await instance.delete(endpoints.delete, {
      params: { user_run_uuid: userRun.uuid },
    });
    return schemas.UserRunSchema.parse(response.data);
  }

  return {
    create: createUserRun,
    getMany: getManyUserRuns,
    get: getUserRun,
    delete: deleteUserRun,
  } as const;
}
